### ISSUE TITLE
Dialogs sometimes crash after beta07

### ISSUE THREAD

**Issue by rjrjr** [square/workflow-kotlin#863]
~I strongly suspect this is a single bug with two symptoms, down to sometimes calling `decorView` before we should.~

Update: actually the placement is likely due to an API change, this looks like the default behavior of the new `onUpdateBounds` param to `OverlayDialogHolder`.

https://workflow-community.slack.com/archives/CHTFPR277/p1660654916735659

> Just a heads up, started seeing this crash after updating to beta09. Happens when we try to show an alert dialog using an AlertOverlay. Not sure exactly how to reproduce it yet:

```
Fatal Exception: android.util.AndroidRuntimeException: requestFeature() must be called before adding content
       at com.android.internal.policy.impl.PhoneWindow.requestFeature(PhoneWindow.java:354)
       at com.android.internal.app.AlertController.installContent(AlertController.java:234)
       at android.app.AlertDialog.onCreate(AlertDialog.java:356)
       at android.app.Dialog.dispatchOnCreate(Dialog.java:373)
       at android.app.Dialog.show(Dialog.java:274)
       at com.squareup.workflow1.ui.container.DialogSession.showDialog(DialogSession.java:94)
       at com.squareup.workflow1.ui.container.LayeredDialogSessions.update(LayeredDialogSessions.kt:163)
       at com.squareup.workflow1.ui.container.BodyAndOverlaysContainer.update(BodyAndOverlaysContainer.java:61)
       at com.squareup.workflow1.ui.container.BodyAndOverlaysContainer$Companion$1$1$1.showRendering(BodyAndOverlaysContainer.kt:150)
       at com.squareup.workflow1.ui.container.BodyAndOverlaysContainer$Companion$1$1$1.showRendering(BodyAndOverlaysContainer.kt:149)
       at com.squareup.workflow1.ui.RealScreenViewHolder.runner$lambda-0(RealScreenViewHolder.java:19)
       at com.squareup.workflow1.ui.RealScreenViewHolder$$InternalSyntheticLambda$2$6d3886a2bf010d8c3ab0c6c37f5168e8d3e5a8edc12d2689a354e9ce9abc4799$0.showRendering(RealScreenViewHolder.java:4)
       at com.squareup.workflow1.ui.container.EnvironmentScreenViewFactoryKt$EnvironmentScreenViewFactory$lambda-2$$inlined$toUnwrappingViewFactory$1$1$1$1$1.invoke(ScreenViewFactory.kt:429)
       at com.squareup.workflow1.ui.container.EnvironmentScreenViewFactoryKt$EnvironmentScreenViewFactory$lambda-2$$inlined$toUnwrappingViewFactory$1$1$1$1$1.invoke(ScreenViewFactory.kt:428)
       at com.squareup.workflow1.ui.container.EnvironmentScreenViewFactoryKt$EnvironmentScreenViewFactory$lambda-2$$inlined$toUnwrappingViewFactory$1$1$1$1.showRendering(ScreenViewFactory.kt:442)
       at com.squareup.workflow1.ui.ScreenViewHolderKt.show(ScreenViewHolderKt.java:74)
       at com.squareup.workflow1.ui.WorkflowViewStub.show(WorkflowViewStub.kt:215)
       at com.squareup.workflow1.ui.WorkflowLayout.show(WorkflowLayout.java:66)
       at com.squareup.workflow1.ui.WorkflowLayout.show$default(WorkflowLayout.java:62)
       at com.squareup.workflow1.ui.WorkflowLayout$take$1$1$1.emit(WorkflowLayout.kt:96)
       at com.squareup.workflow1.ui.WorkflowLayout$take$1$1$1.emit(WorkflowLayout.kt:96)
```

> Some overlays are also misplaced.

beta07            |  beta06
:-------------------------:|:-------------------------:
![](https://user-images.githubusercontent.com/22435681/184931326-d922e926-ac16-4f9a-87f0-8d2acac14cb9.png) |  ![](https://user-images.githubusercontent.com/22435681/184931350-bf86996a-f000-486f-affe-3fe1009c92fa.png)

**Comment by bnvinay92** [comment]
You can place the images side by side with:
Before            |  After
:-------------------------:|:-------------------------:
![](https://user-images.githubusercontent.com/22435681/184931326-d922e926-ac16-4f9a-87f0-8d2acac14cb9.png) |  ![](https://user-images.githubusercontent.com/22435681/184931350-bf86996a-f000-486f-affe-3fe1009c92fa.png)

**Comment by rjrjr** [comment]
Fancy!

**Comment by rjrjr** [comment]
@bnvinay92 The bad placement is actually in your control -- I changed the contract on you. `OverlayDialogHolder()` now has an `onUpdateBounds` argument that defaults to calling `dialog.setBounds`. You'll want to set that to `null`, like `AlertOverlayDialogFactory` does.

Feedback welcome on how to make that more discoverable. Maybe the default should be `null` after all?

**Comment by rjrjr** [comment]
Having trouble reproducing this. I wonder if the non-null `onUpdateBounds` is the cause?

**Comment by rjrjr** [comment]
@bnvinay92 would you be comfortable sharing a gist of your `Dialog` building code? Both the `OverlayDialogFactory` and the `AlertDialog` subclass, if you have one.

**Comment by bnvinay92** [comment]
So the crash was seen only on vanilla alert dialogs which are instances of `AlertOverlay` rendered by `BodyAndOverlaysScreen` from the library. Yes, the placement issue is fixed by setting `onUpdateBounds` to null.

What could be relevant is that these workflows are housed in a fragment that looks like [this](https://gist.github.com/bnvinay92/1f602881fcbe6bafcb9c8939fc7bd67a).

While trying to reproduce this issue, I hardcoded an AlertOverlay in my rendering so that a dialog is always rendered. It does not crash, but what's interesting is that on beta09 it does not render, whereas in beta06 it does.

**Comment by rjrjr** [comment]
I suspect that the crasher is because the first call to `OverlayDialogHolder.show` happens before the first call to `Dialog.show()`, and after beta06 it was changed from this:

```kotlin
  runner(overlay, environment + (InOverlay to overlay))
```

to this:

```kotlin
  dialog.overlay = overlay
  runner(overlay, environment)
```

That `dialog.overlay = overlay` call is calling `window.decorView.setTag()` under the hood. I've learned way too many times that calling `decorView` before `Dialog.show()` is dangerous, especially with `AlertDialog`.

**Comment by rjrjr** [comment]
I moved `InOverlay` out of the `ViewEnvironment` as part of the fix for #844 -- keeping fresh ui models (renderings) in the `ViewEnvironment` was causing `@Composable` functions to recompose on every update, which is intolerable. Can't put it back there; can't safely mess with the decorview to put it in place.

I'm going to look again at what we're actually using the damn thing for, see if we can either get rid or it or at least replace it with something less volatile in the `ViewEnvironment`.

**Comment by rjrjr** [comment]
Answer is: no uses outside of `OverlayDialogHolderKt#canShow`. Maybe we can just hold on to it in the lambda there…
